linked_list/linked_list.py

```diff
@@ -13,67 +12,157 @@ def __init__(self):
+        while current_node is not None:
+            if current_node.value == value:
+                return True
+            current_node = current_node.next
+        return False
+            
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        node_count = 0
+        current_node = self.head
+        while current_node is not None:
+            node_count += 1
+            current_node = current_node.next
+        return node_count
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        current_node = self.head
+        current_index = 0
+
+        if current_node is None:
+            return None
+        
+        while current_node is not None:
+            if current_index == index:
+                return current_node.value
+            else:
+                current_node = current_node.next
+                current_index+=1
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        current_node = self.head
+
+        if current_node is None:
+            return None
+        
+        while True:
+            if current_node.next is None:
+                return current_node.value
+            else:
+                current_node = current_node.next
+
+        
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        new_node = Node(value)
+        current_node = self.head
+
+        if current_node is None:
+            self.head = new_node
+            return
+
+        # What do I want while loop to do? I want while loop to find the last node 
+
+        while current_node is not None:
+            if current_node.next is None:
+                break
+            current_node = current_node.next 
+
+        # When last node is found, I want to append new_node to it.
+        current_node.next = new_node
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        current_node = self.head
+        if current_node is None:
+            return None
+
+        max_value = current_node.value
+
+        while current_node.next:

```

Comment:
I think we can get rid of lines 129-130 if we change the while loop a little:

```suggestion
        while current_node:
```